### DISCUSSION TITLE
Question: Nx's Publishing Opinions

### DISCUSSION THREAD

**Discussion by AndrewOttavianoAbb** [nrwl/nx#8971] [General]
Hello all!

Wasn't sure where to ask a question so I thought here is as good as any. Does Nx have any publishing opinions (specifically for libraries)? What I mean by that is does Nx make assumptions about how the user wants to publish their components from a `--publishable` library. I am looking at tools that can help me build a reuesable component library for our design system that multiple applications can pull from NPM to utilize. As far as I know, there are two main publishing models: **1)** Publish all the components as one large package with one version for the whole thing or **2)** Publish all components individually, each with it's own version. For model 1, take [mui/material-ui](https://github.com/mui/material-ui) (I know, not an Nx repo, but it illustrates the point) as an example. For model 2, take a look at [cerner/terra-core](https://github.com/cerner/terra-core)'s [button package](https://www.npmjs.com/package/terra-button) and its [alert package](https://www.npmjs.com/package/terra-alert); notice that both npm pages point back to the same monorepo in github.

The reason I ask this question here is because I read [this page](https://nx.dev/structure/buildable-and-publishable-libraries) in the Nx documentation, specifically this section (4th paragraph under [Publishable Libraries](https://nx.dev/structure/buildable-and-publishable-libraries#publishable-libraries):
>Keep in mind that the --publishable flag does not enable automatic publishing. Rather it adds to your Nx workspace library a builder target that compiles and bundles your app. _The resulting artifact will be ready to be published to some registry (e.g. [npm](https://npmjs.com/))._

I don't think it's explicitly stated, but _artifact_ (singular) clued me that the tool expects the user to publish a `--publishable` library as one npm package. I would expect a tool as flexible as Nx would have mechanisms to allow one to choose (love that about Nx btw! <3), but I know the way tools are built can impact how consumers build their projects.

Thanks!

**Comment by AndrewOttavianoAbb** [comment]
From @AgentEnder in the [Slack chat](https://nrwlcommunity.slack.com/archives/CMFKWPU6Q/p1645123627481659)

Each published npm package has to have its own package.json file. Looking at the cerna library you sent, that is how they have it setup too, with each publishable component as its own project in lerna.
By splitting things up you get much more benefits out of the dependency graph, caching etc functions from Nx too, so that is what makes sense in most cases.

**Comment by GreatState** [comment]
Now lerna is a dead repo. Any plans to suppport publishing in nx? https://github.com/lerna/lerna

**Reply by AgentEnder** [reply]
We are working on the publishing story, with some minimal support currently active. See: https://nx.dev/getting-started/nx-and-typescript#publish-your-typescript-packages-to-npm

**Reply by subhranshudas** [reply]
Hello @AgentEnder when is the seamless publishing DX coming out? any links or pointers (GH thread)? 
Currently i am trying to create independently publishable packages, but struggling to create a decent script without issues. I have git-cz setup and trying to pick up from affected libs and run publish on them. Any pointers would be greatly appreciated.

**Reply by dylan-westbury** [reply]
hi @AgentEnder , any updates?

**Reply by AgentEnder** [reply]
We are investigating some more native publishing on the Nx side, but lerna's not a dead repo anymore. We (nrwl) took stewardship of it and have released a few major versions over the past year or so.

- https://lerna.js.org/docs/lerna-and-nx

**Comment by AndrewOttavianoAbb** [comment]
Something I also noticed with publishable libraries is that they are built as one large exported module which is not good for library tree shaking, according to this [article](https://blog.theodo.com/2021/04/library-tree-shaking/). That is changeable because there are ways to specify a rollup config for nx to merge into its defaults, but wouldn't it make sense to use the `preserveModules` flag in rollup?

Additionally, that article suggests using the `sideEffects` flag in the package.json so webpack can optimize the library build further (see the [webpack docs](https://webpack.js.org/guides/tree-shaking/#clarifying-tree-shaking-and-sideeffects)). Wouldn't it make sense to add both of these to libraries by default?
